### PR TITLE
Instrument with open tracing

### PR DESCRIPTION
**Pull request by Moocar**
Add tracing to Gatsby! This PR adds [open tracing](http://opentracing.io/) instrumentation to Gatsby so that the following spans are automatically traced:

- Top level build spans
- Bootstrap activities
- API calls (and their implementing plugins)

The traces can be viewed with any open tracing compatible backend by adding the appropriate library dependency and passing its configuration via the new `--open-tracing-config-file` command line option.

Documentation on how to use it is provided in the new `/docs/performance-tracing.md` doc.

Background information on why this approach to tracing was chosen can be found in #1074 

I'm still relatively new to the javascript world, so syntax level feedback would be very much appreciated.

### CODE REVIEW

**Comment by Moocar** [comment]
Note: This shouldn't be merged until https://github.com/DanielMSchmidt/zipkin-javascript-opentracing/pull/68 has been merged.

**Comment by KyleAMathews** [comment]
Fantastic work! So excited about this. Could you post a screenshot or two of some sample traces?

Re) code style — looks good to me. We do use Prettier for formatting so run `npm run format` from the root of the repo (if you haven't) to clean that up.

**Comment by Moocar** [comment]
Cool, added an image in the docs and ran `format`

**Comment by m-allanson** [comment]
I nearly merged this - so I'll repeat [@Moocar's comment above](https://github.com/gatsbyjs/gatsby/pull/6347#issuecomment-403362313) - don't merge this until [this PR has been merged](https://github.com/DanielMSchmidt/zipkin-javascript-opentracing/pull/68).

File: packages/gatsby/src/utils/tracer/index.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  opentracing.initGlobalTracer(tracer)
+
+  return tracer
+}
+
+async function stopTracer() {
+  if (tracerProvider) {
+    if (tracerProvider.stop) {
```
Comment (m-allanson) [inline packages/gatsby/src/utils/tracer/index.js]: This was the only syntax nitpick I could find - you could also write it as `if (tracerProvider && tracerProvider.stop) {`

I'll be over here painting my bikeshed.

**Comment by m-allanson** [comment]
That failed Netlify build is a red herring. Merging, thanks again @Moocar!

**Comment by m-allanson** [comment]
Published as `gatsby@2.0.0-beta.25`

**Comment by jlengstorf** [comment]
Hiya @Moocar! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs

**Pull request merged**